### ISSUE TITLE
Implement functionality to publish Client profiles, Prekey profiles and prekey messages

### ISSUE THREAD

**Issue by claucece** [twstrike/libotrv4#98]
## Why

As we are currently implementing the revision number 2 of the OTRv4 specification, we need to include a consistent way of using publishing user profiles, prekey profiles and prekey messages.

## Tasks
- [x] Create a functionality to publish Client profiles, Prekey profiles and prekey messages

## Depends on

Issue #78 to be finished.

## Open questions

None.

This depends on #108

**Comment by claucece** [comment]
Let's wait on this issue.

**Comment by juniorz** [comment]
I believe the minimum for this that could be done before there is an implementation of a client for the prekey server protocol is:

- [x] Generate a "Prekey Publication Message" (and serialize it).
- [x] Receive a (serialized) "Prekey Ensemble Retrieval Message", deserialize it and validate its prekey ensembles (and filter out invalid ensembles).

As part of #68, the protocol is capable of receiving a prekey ensemble, validates it and send an "offline" data message.

**Comment by juniorz** [comment]
#78 is done, and I believe we could also move this issue to "Todo" when we feel like it.

**Comment by claucece** [comment]
@juniorz OK, let's finish the Non-Interactive-Auth and the heartbeat after the DAKE, and move this to TODO.

**Comment by claucece** [comment]
Note from other issue that is very similar:

For a mode where online and offline are accepted, the same Prekey Server can be used. For a mode with no offline, we need to define how to publish.

I think we can still use the Prekey Servers as the standard for this - but maybe in XMPP exposed under a different service name.

**Comment by claucece** [comment]
@juniorz @giovaneliberato they can be published now?

**Comment by juniorz** [comment]
There is a pretty basic implementation of this. And we are not, for example, filtering out duplicate or invalid ensembles returned by the server.

**Comment by claucece** [comment]
@juniorz ok. Thanks!

**Comment by claucece** [comment]
I'll move the filtering to another issue, as this is mostly closed.
